Jasmine/Grid/ViewModels/CiHuiGridViewModel.swift

```diff
@@ -0,0 +1,54 @@
+import Foundation
+
+class CiHuiGridViewModel: BaseGridViewModel {
+    /// The game data of this game.
+    let gameData: GameData
+
+    init(time: TimeInterval, gameData: GameData, numberOfPhrases: Int) {
+        self.gameData = gameData
+        let phrases = gameData.phrases.next(count: numberOfPhrases)
+
+        var tiles: [String] = []
+        for phrase in phrases {
+            let hanzi = phrase.chinese.characters.map { String($0) }
+            let pinyin = phrase.pinyin.components(separatedBy: " ")
+
+            tiles += (hanzi + pinyin)
+        }
+
+        super.init(time: time, tiles: tiles, rows: numberOfPhrases, columns: Constants.Game.Grid.columns)
+
+        gameTitle = Constants.Game.Grid.CiHui.gameTitle
+        gameInstruction = Constants.Game.Grid.CiHui.gameInstruction
+    }
+
+    /// Returns if and only if the game is won, that is: every row contains a valid Cihui and its pinyin.
+    override var hasGameWon: Bool {
+        for row in 0..<gridData.numRows {
+            var firstText = ""
+            var secondText = ""
+
+            for col in 0...1 {
```

Comment:
maybe divide the num of col by 2 to get the midpoint instead of hardcoding? So that if let's say we do 8x8 mode with 4 chengyu matching 4 pinyin it still works.